src/expr/elim_shadow_converter.h

```diff
@@ -0,0 +1,65 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Andrew Reynolds
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2022 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Implementation of shadow elimination node conversion
+ */
+#include "cvc5_private.h"
+
+#ifndef CVC4__EXPR__ELIM_SHADOW_NODE_CONVERTER_H
+#define CVC4__EXPR__ELIM_SHADOW_NODE_CONVERTER_H
+
+#include "expr/node.h"
+#include "expr/node_converter.h"
+
+namespace cvc5::internal {
+
+/**
+ * This converts a node into one that does not involve shadowing with the
+ * given variables. In particular, if the given vars passed to the constructor
+ * are bound in any quantifier in a subterm of the node to convert, they
+ * are replaced by fresh variables.
+ *
+ * Shadowed variables may be introduced when e.g. quantified formulas
+ * appear on the right hand sides of substitutions in preprocessing. They should
+ * be eliminated by the rewriter. This utility is the standard method for
+ * eliminating them.
+ */
+class ElimShadowNodeConverter : public NodeConverter
+{
+ public:
+  ElimShadowNodeConverter(const Node& q);
+  ~ElimShadowNodeConverter() {}
+  /**
+   * Convert node n as described above during post-order traversal. This
+   * typically should be a subterm of the body of q, assuming that convert
+   * was called on the body of q.
+   */
+  Node postConvert(Node n) override;
+  /**
+   * Get the bound variable used for eliminating shadowing of variable v
+   * bound by closure n that occurs as a subterm of closure q.
+   */
+  static Node getElimShadowVar(const Node& q, const Node& n, const Node& v);
+
+  /** Eliminate shadowing in the closure q */
+  static Node eliminateShadow(const Node& q);
+
+ private:
+  /** The quantified formula to eliminate shadowing from */
+  Node d_quant;
```

Comment:
It's more general than quantifiers. You should update all references to "quant.*" in this class.

Reply:
Done.